### PR TITLE
[cxx-interop] Disambiguate template instantiations with parameter packs

### PR DESCRIPTION
**Pull request by egorzhdan**
This makes sure that different template instantiations of `std::tuple` get distinct Swift type names.

Similar to aa6804a3.

This also refactors `swift::importer::printClassTemplateSpecializationName` to follow a proper visitor pattern for the C++ template arguments.

rdar://139435937

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by Xazax-hun — APPROVED** [review]
LGTM!

File: lib/ClangImporter/ClangClassTemplateNamePrinter.cpp [hunk 1]
```diff
@@ -150,41 +151,106 @@ struct TemplateInstantiationNamePrinter
...
+      if (needsComma)
+        buffer << ", ";
+      Visit(arg, buffer);
+      needsComma = true;
+    }
+  }
+
+  // Unfortunately, clang::templateargumentvisitor::Base does not correctly
+  // perform static dispatch for methods that aren't overloaded, so we have to
+  // overload all of them here.
```
[the remark this answers is not in the document]
Comment (egorzhdan) [reply lib/ClangImporter/ClangClassTemplateNamePrinter.cpp:210]: Patch up: https://github.com/llvm/llvm-project/pull/115336

**Review by slavapestov — COMMENTED** [review]
It seems you're converting a Clang Type into something that looks like a string printout of a Swift Type. Don't you already have the Swift Type for the template argument somewhere so you can just print that out instead?

**Comment by egorzhdan** [comment]
@slavapestov we don't always import the types of template arguments. They are only imported if used elsewhere in the header within fields, parameters, etc. The template parameter could also have a type that we don't import into Swift, or it can be a literal value or an expression that we wouldn't import into Swift.

**Comment by egorzhdan** [comment]
Please test with following PR:

https://github.com/swiftlang/llvm-project/pull/9561

@swift-ci please smoke test

**Comment by egorzhdan** [comment]
Please test with following PR:

https://github.com/swiftlang/llvm-project/pull/9561

@swift-ci please clean smoke test

**Comment by egorzhdan** [comment]
Please test with following PR:

https://github.com/swiftlang/llvm-project/pull/9561

@swift-ci please smoke test

**Comment by egorzhdan** [comment]
Please test with following PR:

https://github.com/swiftlang/llvm-project/pull/9561

@swift-ci please smoke test

**Pull request merged**